prompttools/benchmarks/benchmark.py

```diff
@@ -0,0 +1,97 @@
+        response_options (list(str)): possible responses to measure against
+        correct_response_index (list(int)): list of index of correct response in response_options
+    """
+
+    def __init__(
+        self,
+        experiment: Any,
+        eval_method: Callable,
+        prompts: List[str],
+        response_options: List[Any],
+        correct_response_indices: Optional[List[int]] = None,
+    ):
+        self.experiment = experiment
+        self.eval_method = eval_method
+        self.prompts = prompts
+        self.response_options = response_options
+        self.correct_response_indices = correct_response_indices
+
+    def _get_precision(
+        self,
+        dataframe: pd.DataFrame,
+        pred_col: str,
+        label_col: str,
+    ) -> float:
+        r"""
+        Calculate precision.
+        """
+        # TODO: coming soon
+        pass
+
+    def multiple_choice_accuracy(
+        self,
+        dataframe: pd.DataFrame,
+        col1: str,
+        col2: str,
+    ) -> float:
+        r"""
+        Benchmark LLM accuracy on multiple choice
+        prompt endings.
+        """
+        correct = 0
+        for _, row in dataframe.iterrows():
+            if row[col1] == row[col2]:
+                correct += 1
+        return correct / len(dataframe)
+
+    def multiple_choice_benchmark(
+        self,
+    ) -> Any:
+        r"""
+        Run model experiments to measure response quality.
+        """
+        self.experiment.run()
+
+        if "prompt" not in self.experiment.full_df.columns:
+            # Assume messages column is in place of prompt
+            self.experiment.full_df["prompt"] = self.experiment.full_df["messages"].map(lambda x: str(x))
+            warnings.warn("Column 'prompt' does not exist. Using column 'messages' instead.", UserWarning, stacklevel=2)
+        # Get option with highest similarity to LLM response
+        self.benchmark_df = self.experiment.full_df[["prompt", "response"]]
+        self.benchmark_df["response_options"] = self.response_options
+        self.benchmark_df = self.benchmark_df.explode(column="response_options").reset_index()
+        scores = []
+        for _, row in self.benchmark_df.iterrows():
+            scores.append(self.eval_method(row=row, expected=row["response_options"]))
+        self.benchmark_df["scores"] = scores
+        self.benchmark_df['max_value'] = self.benchmark_df.groupby("prompt")["scores"].transform('max')
+        self.benchmark_df = self.benchmark_df[self.benchmark_df["scores"] == self.benchmark_df["max_value"]]
+        self.benchmark_df = self.benchmark_df.sort_index()
+        # Colect model choices
+        model_choice = []
+        for i, choice in enumerate(self.benchmark_df["response_options"].values):
+            model_choice.append(self.response_options[i].index(choice))
+        self.benchmark_df["model_choice"] = model_choice
+        self.benchmark_df["labels"] = self.correct_response_indices
```

Comment:
Question: Do we plan to return `benchmark_df somewhere?

Reply:
Ah, thanks for pointing that out. My thought process must have veered. Removing self for readability.